doc/design/reader/low_level_reader.md

```diff
@@ -0,0 +1,71 @@
+# Paddle Low Level Data Reader
+
+This design doc will describe a way to implement C++ written data reader(feeder) to
+feed data to paddle when train or inference inorder to improve performance, and make
+best usage of CPU and GPU usage.
+
+This design can be used for both V2 API and new operator design.
+
+## Background
+
+We have sevaral issues talking about training with v2 API get bad training performance,
```

Comment:
I think we need more profiling before reaching the conclusion that v2 API get bad training performance because of data reader. Please take a look at [this comment](https://github.com/PaddlePaddle/Paddle/issues/3675#issuecomment-332996642), we don't have enough evidence that the reader is the problem of the low performance.

That being said, the reader do have two areas of improvement:

1. Use [paddle.reader.buffered](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/reader/decorator.py#L162) decorator to prefetch the data in the background thread. Please see example [here](https://github.com/PaddlePaddle/Paddle/blob/63bc2ff876724016062223bc3c966a4db43ae9c5/doc/design/reader/README.md#prefetch-data)

1. Even if data is prefetched from a background thread, the implementation `paddle.train` does not pre-convert the training data to GPU memory. We can add that support **without** modifying the reader interface that the user uses.

Reply:
That's correct. My first thought was before implementing the reader, benchmark v1 reader and v2 reader by running a training instance with the same dataset and model. Then measure the throughput of training, like `MB/s`.

Reply:
In the past infer jobs, we can get >=80 GPU utilization just used multi-thread reader and a queue just like v2 now. 
1. We need a  benchmark to prove copy from python to c++ is the bottle and
2. In most of the scenes, other takes the longest time not copy data.We have to consider how much will improve after we improve reading speed.
```
后续把用户及视频内容的泛化特征给都去掉，只保留id特征，reader的时间未发生变化（因为读取的数据字段还是跟之前一样，只是在模型训练的时候未用到，所以这块儿的处理时间肯定是一样的），但每个pass的处理时间却快了10倍，由原来的70秒左右变成了7秒左右，具体如下
```